File: singly_linked_list_remove_duplicates/js/SinglyLinkedList.js

function SinglyLinkedListNode(data, next_node) {
  this.data = data;
  this.next_node = next_node;
}

function SinglyLinkedList(items) {
  this.head = null;
  this.tail = null;

  if (typeof items !== 'undefined') {
    for (var i = 0; i < items.length; i++) {
      this.append(items[i]);
    }
  }
}

SinglyLinkedList.prototype.append = function(item) {
  return this.appendNode(new SinglyLinkedListNode(item));
}

SinglyLinkedList.prototype.appendNode = function(node) {
  if (this.tail) {
    this.tail.next_node = node;
    this.tail = node;
  } else {
    this.head = this.tail = node;
  }
  return this;
}

SinglyLinkedList.prototype.insert = function(position, data) {
  if (position == 0 && !this.tail) {
    this.append(data);
  } else {
    var node = this.head;
    var prev = null;
    while (node && position > 0) {
      prev = node;
      node = node.next_node;
      position--;
    }

    if (prev) {
      prev.next_node = new SinglyLinkedListNode(data);
      prev.next_node.next_node = node;
      if (prev == this.tail) {
        this.tail = prev.next_node;
      }
    } else {
      var old_head = this.head;
      this.head = new SinglyLinkedListNode(data);
      this.head.next_node = old_head;
    }
  }
  return this;
}

SinglyLinkedList.prototype.delete = function(data) {
  if (!this.head) {
    return this;
  }

  if (this.head.data == data)
  {
    if (this.head == this.tail) {
      this.head = this.tail = null;
    } else {
      this.head = this.head.next_node;
    }
    return this;
  }

  var node = this.head;

  while (node.next_node) {
    if (node.next_node.data == data) {
      if (node.next_node.next_node == null) {
        this.tail = node;
      }
      node.next_node = node.next_node.next_node;
      return this;
    }
    node = node.next_node;
  }
  return this;
}

module.exports.SinglyLinkedList = SinglyLinkedList;

File: ruby/singly_linked_list/lib/singly_linked_list.rb
require "singly_linked_list/version"

class SinglyLinkedList

  class Node
    attr_accessor :data, :next_node
    def initialize(data = nil, next_node = nil)
      @data = data
      @next_node = next_node
    end
  end

  attr_accessor :head, :tail

  def initialize(items = nil)
    @head = nil
    @tail = nil
    if items
      items.each {|i| append(i) }
    end
  end

  def append data
    append_node(Node.new(data))
  end

  def append_node node
    if @tail
      @tail.next_node = node
    else
      @head = node
    end

    node_tail = node
    cur = node.next_node
    while cur
      node_tail = cur
      cur = cur.next_node
    end

    @tail = node_tail
    return self
  end

  def to_s
    result = "["
    node = @head
    if node
      result << node.data.to_s
      node = node.next_node
      while node
        result << " -> "
        result << node.data.to_s
        node = node.next_node
      end
    end
    result << "]"
    return result
  end

  def insert(position, data)
    if position == 0 && !@tail
      append(data)
    else
      node = @head
      prev = nil
      while node && position > 0
        prev = node
        node = node.next_node
        position -= 1
      end

      if prev
        prev.next_node = Node.new(data)
        prev.next_node.next_node = node
        if prev == @tail
          @tail = prev.next_node
        end
      else
        old_head = @head
        @head = Node.new(data)
        @head.next_node = old_head
      end
    end
    return self
  end

  def delete(data)
    if !@head
      return self
    end

    if @head.data == data
      if @head == @tail
        @head = @tail = nil
      else
        @head = @head.next_node
      end
    else
      node = @head
      while node.next_node
        if node.next_node.data == data
          if !node.next_node.next_node
            @tail = node
          end
          node.next_node = node.next_node.next_node
        else
          node = node.next_node
        end
      end
    end
  end

  include Enumerable

  def each
    node = @head
    while node
      yield node.data
      node = node.next_node
    end
  end

end
